pkg/adapter/apiserver/adapter_test.go

```diff
@@ -17,187 +17,263 @@ limitations under the License.
+					Version:  "v1",
+					Resource: "replicasets",
+				},
+				Controller: true,
+			}},
 		},
 	}
+	for n, tc := range testCases {
+		t.Run(n, func(t *testing.T) {
 
-	for _, tc := range table {
-		t.Run(tc.Name, func(t *gotesting.T) {
-			// Create fake sink server
-			h := &fakeHandler{
-				handler: tc.sink,
-			}
-
-			sinkServer := httptest.NewServer(h)
-			defer sinkServer.Close()
+			a := NewAdaptor(tc.source, k8s, ce, logger, tc.opt)
 
-			// Bind cloud event client
-			ceClient, err := kncloudevents.NewDefaultClient(sinkServer.URL)
-			if err != nil {
-				t.Errorf("cannot create cloud event client: %v", zap.Error(err))
+			got, ok := a.(*adapter)
+			if !ok {
+				t.Errorf("expected NewAdapter to return a *adapter, but did not")
 			}
-
-			// Create fake dynamic client
-			dynamicScheme := runtime.NewScheme()
-			client := fakedynamicclientset.NewSimpleDynamicClient(dynamicScheme, tc.InitialState...)
-
-			stopCh := make(chan struct{})
-			defer close(stopCh)
-
-			tif := &duck.TypedInformerFactory{
-				Client:       client,
-				Type:         &duckv1alpha1.AddressableType{},
-				ResyncPeriod: 1 * time.Second,
-				StopChannel:  stopCh,
+			if diff := cmp.Diff(tc.source, got.source); diff != "" {
+				t.Errorf("unexpected source diff (-want, +got) = %v", diff)
 			}
-
-			_, lister, err := tif.Get(schema.GroupVersionResource{Group: "", Resource: "pods", Version: "v1"})
-			if err != nil {
-				t.Fatalf("Get() = %v", err)
+			if diff := cmp.Diff(tc.wantMode, got.mode); diff != "" {
+				t.Errorf("unexpected mode diff (-want, +got) = %v", diff)
 			}
-
-			opt := reconciler.Options{
-				KubeClientSet: fakekubeclientset.NewSimpleClientset(),
-				Logger:        logtesting.TestLogger(t),
+			if diff := cmp.Diff(tc.wantNamespace, got.namespace); diff != "" {
+				t.Errorf("unexpected namespace diff (-want, +got) = %v", diff)
 			}
-
-			r := &Reconciler{
-				Base:         reconciler.NewBase(opt, controllerAgentName),
-				eventsClient: ceClient,
-				lister:       lister,
+			if diff := cmp.Diff(tc.wantGVRCs, got.gvrcs); diff != "" {
+				t.Errorf("unexpected namespace diff (-want, +got) = %v", diff)
 			}
-			ctx := context.Background()
+		})
+	}
+}
 
-			err = r.Reconcile(ctx, tc.Key)
-			if err != nil {
-				t.Errorf("Expected no error")
-			}
+func TestAdapter_StartRef(t *testing.T) {
+	ce := kncetesting.NewTestClient()
+	logger := zap.NewExample().Sugar()
+	k8s := makeDynamicClient(nil)
+	source := "test-source"
+	opt := Options{
+		Mode:      RefMode,
+		Namespace: "default",
+		GVRCs: []GVRC{{
+			GVR: schema.GroupVersionResource{
+				Group:    "",
+				Version:  "v1",
+				Resource: "pods",
+			},
+		}},
+	}
 
-			if diff := cmp.Diff(tc.data, decode(t, h.body)); diff != "" {
-				t.Errorf("incorrect event (-want, +got): %v", diff)
-			}
-		})
+	a := NewAdaptor(source, k8s, ce, logger, opt)
+
+	var err error
+	stopCh := make(chan struct{})
+	go func() {
+		err = a.Start(stopCh)
+	}()
+	time.Sleep(1 * time.Millisecond)
```

Comment:
is this just to make sure we yield the processor? Perhaps use runtime.GoSched() instead?

Reply:
it turns out the test does not work, I am not sure how to get the cache to do it's thing so I am going to skip the test with a TODO for now.